processing/src/main/java/org/apache/carbondata/processing/newflow/steps/input/InputProcessorStepImpl.java

```diff
@@ -0,0 +1,171 @@
+
+  private static final LogService LOGGER =
+      LogServiceFactory.getLogService(InputProcessorStepImpl.class.getName());
+
+  private GenericParser[] genericParsers;
+
+  private List<Iterator<Object[]>> inputIterators;
+
+  public InputProcessorStepImpl(CarbonDataLoadConfiguration configuration,
+      AbstractDataLoadProcessorStep child, List<Iterator<Object[]>> inputIterators) {
+    super(configuration, child);
+    this.inputIterators = inputIterators;
+  }
+
+  @Override public DataField[] getOutput() {
+    DataField[] fields = configuration.getDataFields();
+    String[] header = configuration.getHeader();
+    DataField[] output = new DataField[fields.length];
+    int k = 0;
+    for (int i = 0; i < header.length; i++) {
+      for (int j = 0; j < fields.length; j++) {
+        if (header[j].equalsIgnoreCase(fields[j].getColumn().getColName())) {
+          output[k++] = fields[j];
+          break;
+        }
+      }
+    }
+    return output;
+  }
+
+  @Override public void initialize() throws CarbonDataLoadingException {
+    DataField[] output = getOutput();
+    genericParsers = new GenericParser[output.length];
+    for (int i = 0; i < genericParsers.length; i++) {
+      genericParsers[i] = CarbonParserFactory.createParser(output[i].getColumn(),
+          (String[]) configuration
+              .getDataLoadProperty(DataLoadProcessorConstants.COMPLEX_DELIMITERS));
+    }
+  }
+
+  private int getNumberOfCores() {
+    int numberOfCores;
+    try {
+      numberOfCores = Integer.parseInt(CarbonProperties.getInstance()
+          .getProperty(CarbonCommonConstants.NUM_CORES_LOADING,
+              CarbonCommonConstants.NUM_CORES_DEFAULT_VAL));
+    } catch (NumberFormatException exc) {
+      numberOfCores = Integer.parseInt(CarbonCommonConstants.NUM_CORES_DEFAULT_VAL);
+    }
+    return numberOfCores;
+  }
+
+  private int getBatchSize() {
+    int batchSize;
+    try {
+      batchSize = Integer.parseInt(configuration
+          .getDataLoadProperty(DataLoadProcessorConstants.DATA_LOAD_BATCH_SIZE,
+              DataLoadProcessorConstants.DATA_LOAD_BATCH_SIZE_DEFAULT).toString());
+    } catch (NumberFormatException exc) {
+      batchSize = Integer.parseInt(DataLoadProcessorConstants.DATA_LOAD_BATCH_SIZE_DEFAULT);
+    }
+    return batchSize;
+  }
+
+  @Override public Iterator<CarbonRowBatch>[] execute() {
+    int batchSize = getBatchSize();
+    List<Iterator<Object[]>>[] readerIterators = partitionInputReaderIterators();
+    Iterator<CarbonRowBatch>[] outIterators = new Iterator[readerIterators.length];
+    for (int i = 0; i < outIterators.length; i++) {
+      outIterators[i] = new InputProcessorIterator(readerIterators[i], genericParsers, batchSize);
+    }
+    return outIterators;
+  }
+
+  private List<Iterator<Object[]>>[] partitionInputReaderIterators() {
+    int numberOfCores = getNumberOfCores();
+    if (inputIterators.size() < numberOfCores) {
+      numberOfCores = inputIterators.size();
```

Comment:
please use a different name (suggest parallelism), the meaning of this variable is `min(inputIterator.size(), numberOfCores)`